vm/bhyve/bhyve.go

```diff
@@ -0,0 +1,329 @@
+	"github.com/google/syzkaller/pkg/osutil"
+	"github.com/google/syzkaller/vm/vmimpl"
+)
+
+func init() {
+	vmimpl.Register("bhyve", ctor, true)
+}
+
+type Config struct {
+	Bridge   string `json:"bridge"`   // name of network bridge device
+	Count    int    `json:"count"`    // number of VMs to use
+	HostIP   string `json:"hostip"`   // VM host IP address
+	Mem      string `json:"mem"`      // amount of VM memory
+	Snapshot string `json:"snapshot"` // ZFS snapshot of workdir
+}
+
+type Pool struct {
+	env *vmimpl.Env
+	cfg *Config
+}
+
+type instance struct {
+	cfg     *Config
+	clone   string
+	tapdev  string
+	image   string
+	debug   bool
+	os      string
+	sshkey  string
+	sshuser string
+	sshhost string
+	merger  *vmimpl.OutputMerger
+	vmName  string
+	bhyve   *exec.Cmd
+	workdir string
+}
+
+var ipRegex = regexp.MustCompile(`bound to (([0-9]+\.){3}[0-9]+) `)
+var tapRegex = regexp.MustCompile(`^tap[0-9]+`)
+
+func ctor(env *vmimpl.Env) (vmimpl.Pool, error) {
+	cfg := &Config{
+		Count: 1,
+		Mem:   "512M",
+	}
+	if err := config.LoadData(env.Config, cfg); err != nil {
+		return nil, fmt.Errorf("failed to parse bhyve vm config: %v", err)
+	}
+	if cfg.Count < 1 || cfg.Count > 128 {
+		return nil, fmt.Errorf("invalid config param count: %v, want [1-128]", cfg.Count)
+	}
+	if env.Debug && cfg.Count > 1 {
+		log.Logf(0, "limiting number of VMs from %v to 1 in debug mode", cfg.Count)
+		cfg.Count = 1
+	}
+	pool := &Pool{
+		cfg: cfg,
+		env: env,
+	}
+	return pool, nil
+}
+
+func (pool *Pool) Count() int {
+	return pool.cfg.Count
+}
+
+func (pool *Pool) Create(workdir string, index int) (vmimpl.Instance, error) {
+	inst := &instance{
+		cfg:     pool.cfg,
+		debug:   pool.env.Debug,
+		image:   pool.env.Image,
+		os:      pool.env.OS,
+		sshkey:  pool.env.SSHKey,
+		sshuser: pool.env.SSHUser,
+		workdir: pool.env.Workdir,
+		vmName:  fmt.Sprintf("syzkaller-%v-%v", pool.env.Name, index),
+	}
+
+	ds := strings.Split(inst.cfg.Snapshot, "@")
+	if len(ds) != 2 {
+		return nil, fmt.Errorf("invalid snapshot name %v", inst.cfg.Snapshot)
+	}
+
+	cloneName := fmt.Sprintf("%v/bhyve%v", ds[0], inst.vmName)
+	osutil.RunCmd(time.Minute, "", "zfs", "destroy", cloneName)
+	cloneArgs := []string{
+		"clone",
+		inst.cfg.Snapshot,
+		cloneName,
+	}
+	if _, err := osutil.RunCmd(time.Minute, "", "zfs", cloneArgs...); err != nil {
+		inst.Close()
+		return nil, err
+	}
+	inst.clone = cloneName
+
+	tapdev, err := osutil.RunCmd(time.Minute, "", "ifconfig", "tap", "create")
+	if err != nil {
+		inst.Close()
+		return nil, err
+	}
+	inst.tapdev = tapRegex.FindString(string(tapdev))
+	if _, err := osutil.RunCmd(time.Minute, "", "ifconfig", inst.cfg.Bridge, "addm", inst.tapdev); err != nil {
+		inst.Close()
+		return nil, err
+	}
+
+	if err := inst.Boot(); err != nil {
+		inst.Close()
+		return nil, err
+	}
+
+	return inst, nil
+}
+
+func (inst *instance) Boot() error {
+	image := fmt.Sprintf("%v/bhyve%v/vm.raw", inst.workdir, inst.vmName)
```

Comment:
Where is `vm.raw` coming from?

Reply:
It was hard-coded by accident.

Reply:
So what should it be? I'm trying to get this running based on the description in the `README.md` and I'm failing. Right now, the image files are not found. So where should they be copied to? Inside the `workdir`?

Reply:
Hmm, there are some undocumented dependencies here. Currently the workdir must be a ZFS data set, and the VM image must be at the root of the data set. However I think it can be made to work without having the image in the workdir itself. I will update the patch shortly.

Reply:
Please try the latest diff. I changed the bhyve backend to handle snapshot creation as well. This is a bit more heavyweight, but it simplifies configuration and is required for CI support anyway. Now you just create a dataset, copy the VM image to it, and specify the dataset name in the syzkaller configuration. The workdir can be in the data set, or not.